src/libstore/daemon.cc

```diff
@@ -944,16 +944,21 @@ void processConnection(
         throw Error("the Nix client version is too old");
 
     auto tunnelLogger = new TunnelLogger(to, clientVersion);
-    auto prevLogger = nix::logger;
+    auto prevLogger = nix::logger.get();
     // FIXME
-    if (!recursive)
-        logger = tunnelLogger;
+    if (!recursive) {
+        // kind of a gross hack but we reset here so that prevLogger remains valid
+        // we make sure to free prevLogger in the Finally below
+        nix::logger.reset();
+        logger = std::unique_ptr<Logger>(tunnelLogger);
+    }
 
     unsigned int opCount = 0;
 
     Finally finally([&]() {
         _isInterrupted = false;
         prevLogger->log(lvlDebug, fmt("%d operations", opCount));
+        delete(prevLogger);
```

Comment:
Just don't start doing that. Use smart pointers all the way if we do it. Keep them "smart" that way it'll cleanup for us.

Reply:
Please see the comment above -- happy to keep discussing it. I couldn't make it work easily.